working/control-flow-collections/feature-specification.md

```diff
@@ -0,0 +1,976 @@
+better expressed using this new syntax and give the user the option to
+automatically change it to the new style. Cases where conditional logic has been
+hoisted all the way out of a collection may be hard to detect since the
+resulting code is pretty imperative.
+
+It should be possible to detect cases like:
+
+*   **Switching out an element using a conditional operator:**
+
+    ```dart
+    [
+      before,
+      condition ? first : second,
+      after
+    ]
+    ```
+
+    Fix:
+
+    ```dart
+    [
+      before,
+      if (condition) first else second,
+      after
+    ]
+    ```
+
+*   **Omitting an element using a conditional operator and `null` filtering:**
+
+    ```dart
+    [
+      before,
+      condition ? first : null,
+      after
+    ].where((e) => e != null).toList()
+    ```
+
+    Fix:
+
+    ```dart
+    [
+      before,
+      if (condition) first,
+      after
+    ]
+    ```
+
+*   **Using `Map.fromIterable()`:**
+
+    ```dart
+    Map.fromIterable(things,
+      key: (e) => someExpression(e),
+      value: (e) => anotherExpression(e)
+    )
+    ```
+
+    Fix:
+
+    ```dart
+    {
+      for (var e in things)
+        someExpression(e): anotherExpression(e)
+    }
+    ```
+
+    This may fall down if the two closures are complex but I think that's rare
+    in practice.
+
+I think we probably *don't* want to blanket apply all of these fixes without
+user intervention. There may be some style preferences or the fix may not
+always succeed. This is a big enough change where having a human validate it is
+a good idea.
+
+## Next Steps
+
+As always, the immediate next step of a proposal is running it past the language
+leads and stakeholders.
+
+### Usability studies
+
+This feature has some good things going for it:
+
+*   It is mostly syntax sugar. A front end should be able to compile this down
+    to existing Dart semantics (with perhaps some extra support needed for `if`
+    in const collections). It shouldn't significantly impact the runtime or
+    backends, so the implementation cost should be relatively low.
+
+*   The semantics are narrow and fairly straightforward. It doesn't interact
+    with the type system in complex ways. It doesn't touch tricky parts of the
+    grammar, calling conventions, or runtime behavior. I think the
+    implementation is fairly low-risk. I don't think we're likely to run into
+    major surprises we didn't anticipate during implementation.
+
+However, the human side of this is less certain. I've tried to make the behavior
+intuitive by piggy-backing on syntax users already understand. But I worry that:
+
+*   Users will find it confusing to see `if` or `for` inside a collection
+    literal.
+
+*   Even after understanding it, users may not *like* the syntax.
+
+*   Users may be unhappy that the syntax doesn't go *far enough*. This feature
+    may lead them to expect, say, `while` to work inside a collection. They may
```

Comment:
This is the first instance of the word "while" in this doc. Is it worth it to speak above as to why only two of the control flow statements are considered here? (No `while`, `do`, or `switch`).

Reply:
Good call. Done.